src/FillArrays.jl

```diff
@@ -265,4 +265,35 @@ Base.:*(a::ZerosVecOrMat, b::AbstractVector) = mult_zeros(a, b)
 Base.:*(a::AbstractVector, b::ZerosVecOrMat) = mult_zeros(a, b)
 Base.:*(a::ZerosVecOrMat, b::ZerosVecOrMat) = mult_zeros(a, b)
 
+
+# Unary = / -.
+Base.:+(a::Zeros) = a
+Base.:-(a::Zeros) = a
+
+# eltypes are the same, so result of promotion is guaranteed to be the same.
+function Base.:+(a::Zeros{T, N}, b::AbstractArray{T, N}) where {T, N}
+    size(a) ≠ size(b) && throw(DimensionMismatch("Incompatible matrix add dimensions."))
+    return b
+end
+Base.:+(a::AbstractArray{T, N}, b::Zeros{T, N}) where {T, N} = b + a
+
+# Mismatched eltypes, so have to be careful to select the appropriate eltype.
+function Base.:+(a::Zeros{T, N}, b::AbstractArray{V, N}) where {T, V, N}
+    size(a) ≠ size(b) && throw(DimensionMismatch("Incompatible matrix add dimensions."))
+    return promote_type(T, V) == V ? b : copy!(similar(b, promote_type(T, V)), b)
```

Comment:
This seems like a bad idea: one expects `0 .+ b` to make a copy of `b` and so not making a copy may cause inconsistent bugs in user code.

If a user really doesn't want a copy, they should use `BLAS.axpy!`.

Reply:
> one expects 0 .+ b to make a copy of b and so not making a copy

Code that assumes a copy is code that will mutate `b`, but that code is already broken by commonly-used immutable data structures, like `b::UnitRange`, `b::UniformScaling`, `b::SVector`, and `b::Zeros`. 

I get what you mean in theory, but I'd be curious to see actual code where this is an issue. Is there precedent in Base one way or the other?

Reply:
If a type is immutable, then there is no difference between a copy and not. The issue here is that someone might do something like this:
```julia
function foo(A, B)
       C = A + B
       C[1] = 2
       C
       end
```
This function is broken for `foo(::Zeros, ::Vector)` using the current definition of `+`, as it will modify the input variable `B`.

Reply:
What I mean is that `foo` is already broken for `foo(SVector(1,2), SVector(3,4))`, or `foo(1:2, 3:4)` because `setindex!` is not defined for those data structures.

My point is that if you're writing generic code (the kind that accepts exotic array types like `Zeros`), you should not rely on mutability, and so it does not matter if `b + 0` returns a copy or not.

Reply:
"Broken" by throwing an error is acceptable. Broken by silently mutating the input is not. 

In every other case, `+(A::AbstractVector, B::Vector)` will not return `B`, and there is no way a user will know or expect  it to return `B` in the special case of `A` being a `Zeros`.

Reply:
Coming back to the problem at hand, @willtebbutt  `copy!(similar(b, promote_type(T, V)), b)` turns a `SVector` into an `MVector` and an `UnitRange` into a `Vector`. Maybe those should be special-cased?  Or perhaps the `copy!` should be turned into `map(x->convert(promote_type(T,V), x), b)`? That works for `SVector`, but not ranges.

Reply:
If it actually makes a difference in practice (which seems unlikely since `copy` is pretty fast), we could add a function `lazy_add`, and a macro (e.g. `@lazy_arithmetic`) that changes calls to `+` to calls to `lazy_add`.

Reply:
It's not the copy I'm worried about, it's the memory allocation. But for my use case, I'm likely to use `SMatrix`, so it's moot.

I'm working on Kalman filters, and I liked the idea that if my fully-general code is passed `transition_matrix = Eye(3)`, then its performance will be the same as if I had written a specialized version of the code assuming an identity transition. It bugs me that I'll pay the cost of a `copy`, but I agree that it's most likely negligible compared to the other operations.

Reply:
My suggestion to write code that is performant would be to write things like:
```julia
A .= B .+ C
```
Then we could add an override like (it'll be more complicated when done properly)
```julia
broadcast!(::typeof(+), dest::AbstractVector, A::Zeros, B::AbstractVector) = dest === B ? B : (dest .= B)
```